### PR TITLE
[ENH] Implement AEAttentionBiGRUNetwork

### PR DESCRIPTION
**Pull request by aadya940**
Implements Attention Bidirectional Gated Recurrent Unit based Auto Encoder .

### CODE REVIEW

**Comment by aadya940** [comment]
@hadifawaz1999 How can we use the Keras Subclassing API in this case, since the tests that don't have tensorflow will fail?

**Comment by MatthewMiddlehurst** [comment]
See the Hydra transformation for ideas, there is ways to get around it. Not sure you can do local classes in Python? Probably dont want to either way 🙂

**Comment by aadya940** [comment]
@MatthewMiddlehurst Thanks :-)

**Comment by aadya940** [comment]
@hadifawaz1999 What will change if we add support for `temporal_latent_space` in this? Do we apply the `Attention` and `self._gate` if `temporal_latent_space` is True since `self._gate` will Flatten the input tensor and convert it from 3 Dimensions to 2 Dimension and the output will no longer be temporal?

**Comment by hadifawaz1999** [comment]
> See the Hydra transformation for ideas, there is ways to get around it. Not sure you can do local classes in Python? Probably dont want to either way 🙂

yup should work like we have in hydra

**Comment by hadifawaz1999** [comment]
> @hadifawaz1999 What will change if we add support for `temporal_latent_space` in this? Do we apply the `Attention` and `self._gate` if `temporal_latent_space` is True since `self._gate` will Flatten the input tensor and convert it from 3 Dimensions to 2 Dimension and the output will no longer be temporal?

let me check the architecture and get back to you

**Comment by hadifawaz1999** [comment]
> > @hadifawaz1999 What will change if we add support for `temporal_latent_space` in this? Do we apply the `Attention` and `self._gate` if `temporal_latent_space` is True since `self._gate` will Flatten the input tensor and convert it from 3 Dimensions to 2 Dimension and the output will no longer be temporal?
> 
> let me check the architecture and get back to you

**Comment by hadifawaz1999** [comment]
if i understand correctly, the number of params for RNN layers depend on the batch size
right ? as it depends on x.shape[0]

**Comment by aadya940** [comment]
Yep, that's what I understood.

**Comment by aadya940** [comment]
@hadifawaz1999 Can we really parametrize the number of layers here? Since,  after the input passes through a Dense layers , its dimensions decrease from 3D to 2D and hence, we cant pass it to the next GRU layer. Also, the implementation here uses just 1 layer in the encoder, as well as the decoder:
https://github.com/blafabregue/TimeSeriesDeepClustering/blob/main/networks/attention_rnn.py

**Comment by aadya940** [comment]
Could we just expand the dims of the tensor after each dense layer as follows?
```
if not (i == self.n_layers_encoder - 1):
                x = tf.expand_dims(x, -1)
```

**Comment by hadifawaz1999** [comment]
should merge main here too @aadya940

**Comment by aadya940** [comment]
#1962

File: aeon/networks/_ae_abgru.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
+"""Implements an Auto-Encoder based on Attention Bidirectional GRUs."""
+
+from aeon.networks.base import BaseDeepLearningNetwork
+
+
```
Comment (hadifawaz1999) [inline aeon/networks/_ae_abgru.py:5]: its missing a reference

**Review by hadifawaz1999 — APPROVED** [review]

**Pull request merged**